File: belfer .NET/dlgPrintPreview.vb
Imports System.Windows.Forms

Public Class dlgPrintPreview
  Public WithEvents Doc As New PrintReport(New DataSet)
  Public Event PreviewModeChange(ByVal IsPreview As Boolean)

  Private Sub OK_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OK_Button.Click
    'Me.DialogResult = System.Windows.Forms.DialogResult.OK
    Dim prnDlg As New PrintDialog
    prnDlg.AllowSomePages = False
    prnDlg.PrinterSettings.FromPage = 1
    prnDlg.PrinterSettings.ToPage = 1
    If prnDlg.ShowDialog = Windows.Forms.DialogResult.OK Then
      'Doc.IsPreview = False
      RaiseEvent PreviewModeChange(False)
      Me.Doc.PageNumber = 0
      'Me.Doc.Offset(0) = 0
      'Me.Doc.Offset(1) = 0
      pvWydruk.Document.DefaultPageSettings.PrinterSettings.Copies = prnDlg.PrinterSettings.Copies
      Me.pvWydruk.Document.DefaultPageSettings.PrinterSettings.FromPage = prnDlg.PrinterSettings.FromPage
      Me.pvWydruk.Document.DefaultPageSettings.PrinterSettings.ToPage = prnDlg.PrinterSettings.ToPage
      Me.pvWydruk.Document.DefaultPageSettings.PrinterSettings.PrinterName = prnDlg.PrinterSettings.PrinterName
      Me.pvWydruk.Document.Print()
      'Doc.IsPreview = True
      RaiseEvent PreviewModeChange(True)
    End If
    'Me.Dispose(True)
  End Sub

  Private Sub Cancel_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Cancel_Button.Click
    Me.DialogResult = System.Windows.Forms.DialogResult.Cancel
    'Me.Dispose(True)
  End Sub

  Private Sub nudZoom_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles nudZoom.ValueChanged
    If Not Me.Created Then Exit Sub
    tbZoom.Value = CType(nudZoom.Value, Integer)
    'Me.pvWydruk.Rows = pvWydruk.Rows
    Me.pvWydruk.Zoom = nudZoom.Value * 0.01
    'Me.Doc.PageNumber = 0
    'Me.Doc.Offset(0) = 0
    'Me.Doc.Offset(1) = 0
    'Me.Doc.Lp = 0
    'Me.pvWydruk.InvalidatePreview()

  End Sub
  Private Sub tbZoom_Scroll(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles tbZoom.Scroll
    'Zoom = tbZoom.Value * 0.01
    nudZoom.Value = tbZoom.Value
    'Me.pvWydruk.Rows = pvWydruk.Rows
    Me.pvWydruk.Zoom = tbZoom.Value * 0.01

  End Sub

  Private Sub dlgPrintPreview_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    pvWydruk.Document = Doc
    pvWydruk.Zoom = nudZoom.Value * 0.01
    pvWydruk.InvalidatePreview()
    'Me.Doc.Offset(0) = 0
    'Me.Doc.Offset(1) = 0
    'Me.Doc.Lp = 0
  End Sub
  Public Sub NewRow()
    pvWydruk.Rows += 1
  End Sub

  Private Sub pvWydruk_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles pvWydruk.Click
    If (Control.ModifierKeys And Keys.Shift) = 0 Then
      If nudZoom.Maximum >= pvWydruk.Zoom * 100 * 2.0 Then
        pvWydruk.Zoom = pvWydruk.Zoom * 2.0
        nudZoom.Value = CType(pvWydruk.Zoom * 100, Decimal)

      End If
    Else
      If nudZoom.Minimum <= pvWydruk.Zoom * 100 / 2.0 Then
        pvWydruk.Zoom = pvWydruk.Zoom / 2.0
        nudZoom.Value = CType(pvWydruk.Zoom * 100, Decimal)

      End If
    End If

  End Sub

  Private Sub cmdPageSetup_Click(sender As Object, e As EventArgs) Handles cmdPageSetup.Click
    Dim pg As New PageSetupDialog, prnVars As New PrintVariables
    pg.EnableMetric = True
    pg.Document = Doc
    pg.MinMargins.Left = CInt(Doc.PrinterSettings.DefaultPageSettings.PrintableArea.Left)
    pg.MinMargins.Top = CInt(Doc.PrinterSettings.DefaultPageSettings.PrintableArea.Top)
    pg.PageSettings.Margins.Top = My.Settings.TopMargin
    pg.PageSettings.Margins.Left = My.Settings.LeftMargin
    pg.PageSettings.Margins.Bottom = My.Settings.TopMargin
    pg.PageSettings.Margins.Right = My.Settings.LeftMargin
    pg.PageSettings.Landscape = My.Settings.Landscape
    pg.AllowPaper = False
    pg.AllowOrientation = True

    If pg.ShowDialog = Windows.Forms.DialogResult.OK Then
      'prnVars.Landscape = pg.PageSettings.Landscape
      'prnVars.LeftMargin = pg.PageSettings.Margins.Left
      'prnVars.TopMargin = pg.PageSettings.Margins.Top
      'prnVars.RightMargin = pg.PageSettings.Margins.Right
      'prnVars.BottomMargin = pg.PageSettings.Margins.Bottom
      My.Settings.TopMargin = pg.PageSettings.Margins.Top
      My.Settings.LeftMargin = pg.PageSettings.Margins.Left
      My.Settings.Landscape = pg.PageSettings.Landscape

      pvWydruk.Rows = 1
      Doc.PageNumber = 0
      'Me.Doc.Offset(0) = 0
      'Me.Doc.Offset(1) = 0
      'Me.Doc.Lp = 0
      pvWydruk.InvalidatePreview()
    End If
  End Sub

  Private Sub cmdFontSetup_Click(sender As Object, e As EventArgs) Handles cmdFontSetup.Click
    Dim fg As New FontDialog, prnVars As New PrintVariables
    fg.MinSize = 8
    fg.MaxSize = 20
    fg.FontMustExist = True
    fg.Font = My.Settings.TextFont
    If fg.ShowDialog = Windows.Forms.DialogResult.OK Then
      'TextFont = fg.Font
      My.Settings.TextFont = fg.Font
      My.Settings.SubHeaderFont = New Font(My.Settings.SubHeaderFont.FontFamily, My.Settings.TextFont.Size + 1, FontStyle.Bold)
      My.Settings.HeaderFont = New Font(My.Settings.HeaderFont.FontFamily, My.Settings.TextFont.Size + 2, FontStyle.Bold)
      pvWydruk.Rows = 1
      Doc.PageNumber = 0
      'Me.Doc.Offset(0) = 0
      'Me.Doc.Offset(1) = 0
      'Me.Doc.Lp = 0
      pvWydruk.InvalidatePreview()
    End If
  End Sub
End Class

File: BelferCS/dlgPrintPreview.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using System.Drawing.Printing;
using BelferCS.Properties;
namespace BelferCS
{
    public partial class dlgPrintPreview : Form
    {
        public dlgPrintPreview()
        {
            InitializeComponent();
            pvWydruk.Document = Doc;
            pvWydruk.Zoom = (double)nudZoom.Value * 0.01;
            GetPageSettings();
            GetFontSettings();
            SetMargins();
        }
        public PrintDocument Doc = new PrintDocument();
        public delegate void PreviewModeDelegate(bool IsPreview);
        public event PreviewModeDelegate PreviewModeChanged;

        private void OK_Button_Click(object sender, EventArgs e)
        {
            var dlgPrint = new PrintDialog();
            dlgPrint.AllowSomePages = false;
            dlgPrint.PrinterSettings.FromPage = 1;
            dlgPrint.PrinterSettings.ToPage = 1;
            if (dlgPrint.ShowDialog() == DialogResult.OK)
            {
                PreviewModeChanged?.Invoke(false);
                pvWydruk.Document.DefaultPageSettings.PrinterSettings.Copies = dlgPrint.PrinterSettings.Copies;
                pvWydruk.Document.DefaultPageSettings.PrinterSettings.FromPage = dlgPrint.PrinterSettings.FromPage;
                pvWydruk.Document.DefaultPageSettings.PrinterSettings.ToPage = dlgPrint.PrinterSettings.ToPage;
                pvWydruk.Document.DefaultPageSettings.PrinterSettings.PrinterName = dlgPrint.PrinterSettings.PrinterName;
                pvWydruk.Document.Print();
                PreviewModeChanged?.Invoke(true);                
            }
        }

        private void Cancel_Button_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.Cancel;
        }

        private void GetPageSettings()
        {
            if (Settings.Default.Landscape) { rbHorizontal.Checked = true; } else { rbVertical.Checked = true; }
            var M = new CalcHelper.Math();
            nudLeftMargin.Value = (decimal)Math.Round(M.INtoMM(Settings.Default.LeftMargin));
            nudTopMargin.Value = (decimal)Math.Round(M.INtoMM(Settings.Default.TopMargin));
        }

        private void GetFontSettings()
        {
            txtStyle.Font = Settings.Default.TextFont;
            txtStyle.Text = Settings.Default.TextFont.FontFamily.Name+"; ";
            txtStyle.Text += Settings.Default.TextFont.Size+"pt; ";
            if (Settings.Default.TextFont.Bold) { txtStyle.Text += "pogrubienie,"; }
            if (Settings.Default.TextFont.Italic) { txtStyle.Text += "kursywa"; }

            txtSHStyle.Font = Settings.Default.SubHeaderFont;
            txtSHStyle.Text = Settings.Default.SubHeaderFont.FontFamily.Name+"; ";
            txtSHStyle.Text += Settings.Default.SubHeaderFont.Size+"pt; ";
            if (Settings.Default.SubHeaderFont.Bold) { txtSHStyle.Text += "pogrubienie,"; }
            if (Settings.Default.SubHeaderFont.Italic) { txtSHStyle.Text += "kursywa"; }

            txtHStyle.Font = Settings.Default.HeaderFont;
            txtHStyle.Text = Settings.Default.HeaderFont.FontFamily.Name+"; ";
            txtHStyle.Text += Settings.Default.HeaderFont.Size+"pt; ";
            if (Settings.Default.HeaderFont.Bold) { txtHStyle.Text += "pogrubienie,"; }
            if (Settings.Default.HeaderFont.Italic) { txtHStyle.Text += "kursywa"; }
        }

        private void nudZoom_ValueChanged(object sender, EventArgs e)
        {
            tbZoom.Value = (int)nudZoom.Value;
            pvWydruk.Zoom = (double)nudZoom.Value * 0.01;
        }

        private void tbZoom_Scroll(object sender, EventArgs e)
        {
            nudZoom.Value = tbZoom.Value;
            pvWydruk.Zoom = tbZoom.Value * 0.01;
        }
        public void NewRow()
        {
            pvWydruk.Rows += 1;
        }

        private void pvWydruk_Click(object sender, EventArgs e)
        {
            if (ModifierKeys == Keys.Shift || ModifierKeys == Keys.Control)
            {
                if ((double)nudZoom.Minimum <= pvWydruk.Zoom * 100 / 2)
                {
                    pvWydruk.Zoom = pvWydruk.Zoom / 2.0;
                    nudZoom.Value = (decimal)pvWydruk.Zoom * 100;
                }                
            }
            else
            {
                if ((double)nudZoom.Maximum >= pvWydruk.Zoom * 100 * 2)
                {
                    pvWydruk.Zoom = pvWydruk.Zoom * 2.0;
                    nudZoom.Value = (decimal)pvWydruk.Zoom * 100;
                }
            }
        }

        private void cmdFontSetup_Click(object sender, EventArgs e)
        {
            var fd = new FontDialog();
            fd.FontMustExist = true;
            fd.Font = Settings.Default.TextFont;
            fd.MinSize = 8;
            fd.MaxSize = 16;
            if (fd.ShowDialog() == DialogResult.OK)
            {
                Settings.Default.TextFont = fd.Font;
                Settings.Default.SubHeaderFont = new Font(Settings.Default.SubHeaderFont.FontFamily, Settings.Default.TextFont.Size + 1, FontStyle.Bold);
                Settings.Default.HeaderFont = new Font(Settings.Default.HeaderFont.FontFamily, Settings.Default.TextFont.Size + 2, FontStyle.Bold);
                Settings.Default.Save();
                GetFontSettings();
                pvWydruk.Rows = 1;
                pvWydruk.InvalidatePreview();
            }
        }

        private void rbVertical_CheckedChanged(object sender, EventArgs e)
        {
            if (!((RadioButton)sender).Checked) return;
            Settings.Default.Landscape = ((RadioButton)sender).Name == rbHorizontal.Name ? ((RadioButton)sender).Checked : !((RadioButton)sender).Checked;
            Settings.Default.Save();
            pvWydruk.Document.DefaultPageSettings.Landscape = Settings.Default.Landscape;
            pvWydruk.InvalidatePreview();
        }

        private void nudLeftMargin_ValueChanged(object sender, EventArgs e)
        {
            var CHM = new CalcHelper.Math();
            Settings.Default.LeftMargin = (int)CHM.MMtoIN(((float)((NumericUpDown)sender).Value));
            Settings.Default.Save();
            SetHorizontalMargins();
            pvWydruk.InvalidatePreview();
        }

        private void nudTopMargin_ValueChanged(object sender, EventArgs e)
        {
            var CHM = new CalcHelper.Math();
            Settings.Default.TopMargin = (int)CHM.MMtoIN(((float)((NumericUpDown)sender).Value));
            Settings.Default.Save();
            SetVerticalMargins();
            pvWydruk.InvalidatePreview();
        }
        private void SetMargins()
        {
            SetHorizontalMargins();
            SetHorizontalMargins();
        }
        private void SetHorizontalMargins()
        {
            Doc.DefaultPageSettings.Margins.Left = Settings.Default.LeftMargin;
            Doc.DefaultPageSettings.Margins.Right = Settings.Default.LeftMargin;
        }
        private void SetVerticalMargins()
        {
            Doc.DefaultPageSettings.Margins.Top = Settings.Default.TopMargin;
            Doc.DefaultPageSettings.Margins.Bottom = Settings.Default.TopMargin;
        }
    }
}
